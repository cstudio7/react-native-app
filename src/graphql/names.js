import gql from 'graphql-tag';

export const femaleNamesQuery = gql`
  {
    femaleNames {
      id
      name
    }
  }
`;

export const maleNamesQuery = gql`
  {
    maleNames {
      id
      name
    }
  }
`;
