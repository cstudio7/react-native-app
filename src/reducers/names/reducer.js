const R = require('ramda');

import {
  LISTSCREEN_NAME_FAVORITE,
  LISTSCREEN_NAME_UNFAVORITE,
  FAVSCREEN_NAME_UNFAVORITE,
  FETCH_NAMES_SUCCESS
} from '../../constants/ActionTypes';

const initialState = {
  female: [],
  male: []
};

const updateItemInArray = (state, name, gender) => {
  const genderNames = state[gender];
  const index = R.findIndex(R.propEq('name', name.name))(genderNames);

  return {
    ...state,
    [gender]: [
      ...genderNames.slice(0, index),
      name,
      ...genderNames.slice(index + 1)
    ]
  };
};

const mergeGenderNamesWithState = (state, genderNames = [], gender) => {
  if (!R.path([gender, 'length'], state)) {
    return genderNames;
  }

  return genderNames.map(name =>
    R.mergeAll([
      name,
      state[gender].find(stateName => stateName.name === name.name)
    ])
  );
};

const saveNames = (state, action) => {
  const { response } = action;
  const { female, male } = response;
  return {
    female: mergeGenderNamesWithState(state, female, 'female'),
    male: mergeGenderNamesWithState(state, male, 'male')
  };
};

const favoriteName = (state, action) => {
  const { payload } = action;
  return updateItemInArray(state, payload.name, payload.gender);
};

export default (state = initialState, action) => {
  switch (action.type) {
    case FETCH_NAMES_SUCCESS:
      return saveNames(state, action);
    case LISTSCREEN_NAME_FAVORITE:
    case LISTSCREEN_NAME_UNFAVORITE:
    case FAVSCREEN_NAME_UNFAVORITE:
      return favoriteName(state,action);
    case 'decrement':
      return {count: state.count - 1};
    default:
      // A reducer must always return a valid state.
      // Alternatively you can throw an error if an invalid action is dispatched.
      return state;
  }
}
