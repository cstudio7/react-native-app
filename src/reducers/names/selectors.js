export const getNames = (state, { route, screen }) => {
  const names = state.names[route.key];
  if (screen === 'Favorites') {
    return names.filter(name => name.isFavorite);
  }
  return names;
};