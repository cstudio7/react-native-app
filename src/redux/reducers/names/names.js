const R = require('ramda');

import Actions from '../../../constants/Actions';

export const initialState = {
  female: [],
  male: []
};

const updateName = (name, gender, state) => {
  const genderData = state[gender];
  const index = R.findIndex(R.propEq('id', name.id))(genderData);

  return {
    ...state,
    [gender]: [...genderData.slice(0, index), name, ...genderData.slice(index + 1)]
  };
};

export default (state = initialState, action) => {
  const { type, gender } = action;

  if (type === Actions.saveNames) {
    return {
      ...state,
      [gender]: R.uniq([...state[gender], ...action.names])
    };
  }

  if (type === Actions.favoriteName) {
    return updateName({ ...action.name, isFavorite: true }, gender, state);
  }

  if (type === Actions.unfavoriteName) {
    return updateName({ ...action.name, isFavorite: false }, gender, state);
  }

  return state;
};
