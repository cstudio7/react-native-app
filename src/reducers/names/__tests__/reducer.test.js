import names from '../reducer';
import {
  LISTSCREEN_NAME_FAVORITE,
  FAVSCREEN_NAME_FAVORITE,
  LISTSCREEN_NAME_UNFAVORITE,
  FAVSCREEN_NAME_UNFAVORITE,
  FETCH_NAMES_SUCCESS
} from '../../../constants/ActionTypes';

it('should return the initial state', () => {
  expect(names(undefined, {})).toMatchSnapshot();
});

describe('FETCH_NAMES_SUCCESS', () => {
  it('updates female names', () => {
    const initialState = {
      female: [],
      male: []
    };

    expect(
      names(initialState, {
        type: FETCH_NAMES_SUCCESS,
        response: {
          femaleNames: [{ id: 1, name: 'sveta', isFavorite: true }]
        }
      })
    ).toMatchSnapshot();
  });

  it('updates male names', () => {
    const initialState = {
      female: [],
      male: []
    };

    expect(
      names(initialState, {
        type: FETCH_NAMES_SUCCESS,
        response: {
          maleNames: [{ id: 1, name: 'artur', isFavorite: true }]
        }
      })
    ).toMatchSnapshot();
  });

  it('updates state retrieved from a storage', () => {
    const initialState = {
      female: [],
      male: [
        { id: 1, name: 'artur', isFavorite: true },
        { id: 2, name: 'alex' }
      ]
    };

    expect(
      names(initialState, {
        type: FETCH_NAMES_SUCCESS,
        response: {
          maleNames: [
            { id: 1, name: 'artur', meaning: 'bla' },
            { id: 2, name: 'alex', meaning: 'bla bla' },
            { id: 3, name: 'pierre', meaning: 'bla bla bla' }
          ]
        }
      })
    ).toMatchSnapshot();
  });
});

it('should handle LISTSCREEN_NAME_FAVORITE action', () => {
  const initialState = {
    female: [{ id: 1, name: 'sveta', isFavorite: true }],
    male: [{ id: 1, name: 'artur', isFavorite: false }]
  };

  expect(
    names(initialState, {
      type: LISTSCREEN_NAME_FAVORITE,
      payload: {
        gender: 'male',
        name: { id: 1, name: 'artur', isFavorite: true }
      }
    })
  ).toMatchSnapshot();
});

it('should handle FAVSCREEN_NAME_FAVORITE action', () => {
  const initialState = {
    female: [{ id: 1, name: 'sveta', isFavorite: true }],
    male: [{ id: 1, name: 'artur', isFavorite: false }]
  };

  expect(
    names(initialState, {
      type: FAVSCREEN_NAME_FAVORITE,
      payload: {
        gender: 'male',
        name: { id: 1, name: 'artur', isFavorite: true }
      }
    })
  ).toMatchSnapshot();
});

it('should handle LISTSCREEN_NAME_UNFAVORITE action', () => {
  const initialState = {
    female: [],
    male: [{ id: 1, name: 'artur', isFavorite: true }]
  };

  expect(
    names(initialState, {
      type: LISTSCREEN_NAME_UNFAVORITE,
      payload: {
        gender: 'male',
        name: { id: 1, name: 'artur', isFavorite: false }
      }
    })
  ).toMatchSnapshot();
});

it('should handle FAVSCREEN_NAME_UNFAVORITE action', () => {
  const initialState = {
    female: [],
    male: [{ id: 1, name: 'artur', isFavorite: true }]
  };

  expect(
    names(initialState, {
      type: FAVSCREEN_NAME_UNFAVORITE,
      payload: {
        gender: 'male',
        name: { id: 1, name: 'artur', isFavorite: false }
      }
    })
  ).toMatchSnapshot();
});
