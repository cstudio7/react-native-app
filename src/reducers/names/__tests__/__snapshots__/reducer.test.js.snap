// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FETCH_NAMES_SUCCESS updates female names 1`] = `
Object {
  "female": Array [
    Object {
      "id": 1,
      "isFavorite": true,
      "name": "sveta",
    },
  ],
  "male": Array [],
}
`;

exports[`FETCH_NAMES_SUCCESS updates male names 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "id": 1,
      "isFavorite": true,
      "name": "artur",
    },
  ],
}
`;

exports[`FETCH_NAMES_SUCCESS updates state retrieved from a storage 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "id": 1,
      "isFavorite": true,
      "meaning": "bla",
      "name": "artur",
    },
    Object {
      "id": 2,
      "meaning": "bla bla",
      "name": "alex",
    },
    Object {
      "id": 3,
      "meaning": "bla bla bla",
      "name": "pierre",
    },
  ],
}
`;

exports[`should handle FAVSCREEN_NAME_UNFAVORITE action 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "id": 1,
      "isFavorite": false,
      "name": "artur",
    },
  ],
}
`;

exports[`should handle LISTSCREEN_NAME_FAVORITE action 1`] = `
Object {
  "female": Array [
    Object {
      "id": 1,
      "isFavorite": true,
      "name": "sveta",
    },
  ],
  "male": Array [
    Object {
      "id": 1,
      "isFavorite": true,
      "name": "artur",
    },
  ],
}
`;

exports[`should handle LISTSCREEN_NAME_UNFAVORITE action 1`] = `
Object {
  "female": Array [],
  "male": Array [
    Object {
      "id": 1,
      "isFavorite": false,
      "name": "artur",
    },
  ],
}
`;

exports[`should return the initial state 1`] = `
Object {
  "female": Array [],
  "male": Array [],
}
`;
